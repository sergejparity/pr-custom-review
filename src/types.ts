import github from "@actions/github"
import { RestEndpointMethodTypes } from "@octokit/plugin-rest-endpoint-methods"

export type CommitState =
  RestEndpointMethodTypes["repos"]["createCommitStatus"]["parameters"]["state"]

export type Octokit = ReturnType<typeof github.getOctokit>

export type PR = {
  number: number
  base: {
    repo: {
      name: string
      owner: {
        login: string
      }
    }
  }
  head: {
    sha: string
  }
  user: {
    login: string
  }
  diff_url: string
  html_url: string
}

export type BaseRule = {
  name: string
  condition: string
  check_type: "diff" | "changed_files"
}

export type RuleCriteria = {
  min_approvals: number
  users?: Array<string> | null
  teams?: Array<string> | null
}

export type BasicRule = BaseRule & RuleCriteria

export type OrRule = BaseRule & {
  any: RuleCriteria[]
}

export type AndRule = BaseRule & {
  all: RuleCriteria[]
}

export type RuleKind = "BasicRule" | "OrRule" | "AndRule"
export type Rule = BasicRule | OrRule | AndRule

export type RulesConfigurations = {
  basic: {
    kind: "BasicRule"
    uniqueFields: ["min_approvals", "teams", "users"]
    invalidFields: ["any", "all"]
  }
  and: {
    kind: "AndRule"
    uniqueFields: ["all"]
    invalidFields: ["min_approvals", "teams", "users", "any"]
  }
  or: {
    kind: "OrRule"
    uniqueFields: ["any"]
    invalidFields: ["min_approvals", "teams", "users", "all"]
  }
}

export type Configuration = {
  rules: Rule[]
}

export type RuleUserInfo = { team: string | null }

export type MatchedRule = {
  name: string
  min_approvals: number
  users: Map<string, RuleUserInfo>
  kind: RuleKind
  id: number
}

export class RuleSuccess {
  constructor(public rule: MatchedRule) {}
}
export class RuleFailure {
  constructor(
    public rule: MatchedRule,
    public problem: string,
    public usersToAskForReview: Map<string, RuleUserInfo>,
  ) {}
}
