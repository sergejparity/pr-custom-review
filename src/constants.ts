import { CommitState, RulesConfigurations } from "./types"

export const commitStateSuccess: CommitState = "success"
export const commitStateFailure: CommitState = "failure"

export const rulesConfigurations: RulesConfigurations = {
  basic: {
    kind: "BasicRule",
    uniqueFields: ["min_approvals", "teams", "users"],
    invalidFields: ["any", "all"],
  },
  and: {
    kind: "AndRule",
    uniqueFields: ["all"],
    invalidFields: ["min_approvals", "teams", "users", "any"],
  },
  or: {
    kind: "OrRule",
    uniqueFields: ["any"],
    invalidFields: ["min_approvals", "teams", "users", "all"],
  },
}

export const variableNameToActionInputName = {
  teamLeadsTeam: "team-leads-team",
  locksReviewTeam: "locks-review-team",
}
