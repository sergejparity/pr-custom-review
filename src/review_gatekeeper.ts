export interface Settings {
  approvals: {
    minimum?: number
    groups?: {
      minimum?: number
      name: string
      from: {
        person?: string[]
      }
    }[]
  }
}

function set_equal<T>(as: Set<T>, bs: Set<T>): boolean {
  if (as.size !== bs.size) {
    return false
  }
  for (const a of as) {
    if (!bs.has(a)) {
      return false
    }
  }
  return true
}

function set_intersect<T>(as: Set<T>, bs: Set<T>): Set<T> {
  return new Set([...as].filter(e => bs.has(e)))
}

function set_to_string<T>(as: Set<T>): string {
  return [...as].join(', ')
}

export class ReviewGatekeeper {
  private messages: string[]
  private meet_criteria: boolean

  constructor(settings: Settings, approved_users: string[], pr_owner: string) {
    this.messages = []
    this.meet_criteria = true

    const approvals = settings.approvals
    // check if the minimum criteria is met.
    if (approvals.minimum) {
      if (approvals.minimum > approved_users.length) {
        this.meet_criteria = false
        this.messages.push(
          `${approvals.minimum} reviewers should approve this PR (currently: ${approved_users.length})`
        )
      }
    }

    // check if the groups criteria is met.
    const approved = new Set(approved_users)
    if (approvals.groups) {
      for (const group of approvals.groups) {
        const required_users = new Set(group.from.person)
        // Remove PR owner from required uesrs because PR owner cannot approve their own PR.
        required_users.delete(pr_owner)
        const approved_from_this_group = set_intersect(required_users, approved)
        const minimum_of_group = group.minimum
        if (minimum_of_group) {
          if (minimum_of_group > approved_from_this_group.size) {
            this.meet_criteria = false
            this.messages.push(
              `${minimum_of_group} reviewers from the group '${
                group.name
              }' (${set_to_string(
                required_users
              )}) should approve this PR (currently: ${
                approved_from_this_group.size
              })`
            )
          }
        } else {
          // If no `minimum` option is specified, approval from all is required.
          if (!set_equal(approved_from_this_group, required_users)) {
            this.meet_criteria = false
            this.messages.push(
              `All of the reviewers from the group '${
                group.name
              }' (${set_to_string(required_users)}) should approve this PR`
            )
          }
        }
      }
    }
  }

  satisfy(): boolean {
    return this.meet_criteria
  }

  getMessages(): string[] {
    return this.messages
  }
}
