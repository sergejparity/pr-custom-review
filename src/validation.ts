import Joi from "joi"

import { Configuration, Rule, RuleCriteria } from "./types"

const ruleCriteria = function ({
  isMinApprovalsOptional,
}: {
  isMinApprovalsOptional: boolean
}) {
  let minApprovals = Joi.number().min(1)
  if (isMinApprovalsOptional) {
    minApprovals = minApprovals.optional().allow(null)
  }
  return {
    min_approvals: minApprovals,
    users: Joi.array().items(Joi.string()).optional().allow(null),
    teams: Joi.array().items(Joi.string()).optional().allow(null),
  }
}

const ruleSchema = Joi.object<Rule>().keys({
  name: Joi.string().required(),
  condition: Joi.string().required(),
  check_type: Joi.string().valid("diff", "changed_files").required(),
  ...ruleCriteria({ isMinApprovalsOptional: true }),
  all: Joi.array()
    .items(
      Joi.object<RuleCriteria>().keys(
        ruleCriteria({ isMinApprovalsOptional: false }),
      ),
    )
    .optional()
    .allow(null),
  any: Joi.array()
    .items(
      Joi.object<RuleCriteria>().keys(
        ruleCriteria({ isMinApprovalsOptional: false }),
      ),
    )
    .optional()
    .allow(null),
})

export const configurationSchema = Joi.object<Configuration>().keys({
  rules: Joi.array().items(ruleSchema).required(),
})
