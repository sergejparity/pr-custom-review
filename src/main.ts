import * as core from '@actions/core'
import * as github from '@actions/github'
import * as Webhooks from '@octokit/webhooks-types'
import * as fs from 'fs'
import * as YAML from 'yaml'
import { EOL } from 'os'
import { Settings, ReviewGatekeeper } from './review_gatekeeper'
import { stderr } from 'process'

export async function assignReviewers(client: any, reviewer_persons: string[], reviewer_teams: string[], pr_number: number) {
  try {
    console.log(`entering assignReviewers`) //DEBUG
    console.log(`persons length: ${reviewer_persons.length} - ${reviewer_persons}`) //DEBUG
    if (reviewer_persons) {
      await client.rest.pulls.requestReviewers({
        owner: github.context.repo.owner,
        repo: github.context.repo.repo,
        pull_number: pr_number,
        reviewers: reviewer_persons,
      });
      core.info(`Requested review from users: ${reviewer_persons}.`);
    }
    console.log(`passed by persons trying teams`) //DEBUG
    console.log(`teams length: ${reviewer_teams}`) //DEBUG
    if (reviewer_teams) {
      await client.rest.pulls.requestReviewers({
        owner: github.context.repo.owner,
        repo: github.context.repo.repo,
        pull_number: pr_number,
        team_reviewers: reviewer_teams,
      });
      core.info(`Requested review from teams: ${reviewer_teams}.`);
    }
    console.log(`exiting assignReviewers`) //DEBUG
  } catch (error) {
    core.setFailed(error.message)
    console.log("error: ", error)
  }
}

async function run(): Promise<void> {
  try {
    const context = github.context
    if (
      context.eventName !== 'pull_request' &&
      context.eventName !== 'pull_request_review'
    ) {
      core.setFailed(
        `Invalid event: ${context.eventName}. This action should be triggered on pull_request and pull_request_review`
      )
      return
    }
    const payload = context.payload as
      | Webhooks.PullRequestEvent
      | Webhooks.PullRequestReviewEvent

    const token: string = core.getInput('token')
    const octokit = github.getOctokit(token)
    const repo = payload.repository.url
    const pr_number = payload.pull_request.number
    const pr_diff = payload.pull_request.diff_url
    const pr_owner = payload.pull_request.user.login
    const sha = payload.pull_request.head.sha
    const workflow_url = `${process.env['GITHUB_SERVER_URL']}/${process.env['GITHUB_REPOSITORY']}/actions/runs/${process.env['GITHUB_RUN_ID']}`
    const workflow_name = `${process.env.GITHUB_WORKFLOW}`
    const organization: string = process.env.GITHUB_REPOSITORY?.split("/")[0]!

    console.log(`repo: ${repo}`)
    console.log(`pr_owner: ${pr_owner}`)
    console.log(`diff url: ${pr_diff}`)

    const diff_body = await octokit.request(pr_diff)
    console.log(diff_body.data)

    // experiment with shell exec
    // const { exec } = require("child_process");

    // exec("git --no-pager diff ${{ github.event.pull_request.base.sha }}...${{ github.event.pull_request.head.sha }} -U1 | { grep ðŸ”’ || true; }", (error, stdout, stderr) => {
    //   if (error) {
    //     console.log(`error: ${error.message}`);
    //     return;
    //   }
    //   if (stderr) {
    //     console.log(`stderr: ${stderr}`);
    //     return;
    //   }
    //   console.log(`stdout: ${stdout}`);
    // });


    // No breaking changes - no cry. Set status OK and exit.
    if (process.env.CUSTOM_REVIEW_REQUIRED == 'not_required') {
      console.log(`Special approval of this PR is not required.`)

      octokit.rest.repos.createCommitStatus({
        ...context.repo,
        sha,
        state: 'success',
        context: workflow_name,
        target_url: workflow_url,
        description: "Special approval of this PR is not required."
      })
      return
    }

    // Read values from config file if it exists
    const config_file = fs.readFileSync(core.getInput('config-file'), 'utf8')

    // Parse contents of config file into variable
    const config_file_contents = YAML.parse(config_file)

    const reviewer_persons_set: Set<string> = new Set()
    const reviewer_teams_set: Set<string> = new Set()

    for (const reviewers of config_file_contents.approvals.groups) {
      if (reviewers.from.persons) {
        for (var entry of reviewers.from.persons) {
          if (entry != pr_owner) {
            reviewer_persons_set.add(entry)
          }
        }
      }
      if (reviewers.from.teams) {
        for (var entry of reviewers.from.teams) {
          reviewer_teams_set.add(entry)
        }
      }
    }

    console.log(`persons set:`) //DEBUG
    console.log(reviewer_persons_set) //DEBUG
    console.log(`teams set:`) //DEBUG
    console.log(reviewer_teams_set) //DEBUG
    console.log(Array.from(reviewer_persons_set))  //DEBUG

    if (context.eventName == 'pull_request') {
      console.log(`I'm going to request someones approval!!!`) //DEBUG
      assignReviewers(octokit, Array.from(reviewer_persons_set), Array.from(reviewer_teams_set), pr_number)

      octokit.rest.repos.createCommitStatus({
        ...context.repo,
        sha,
        state: 'failure',
        context: workflow_name,
        target_url: workflow_url,
        description: `PR contains changes subject to special review. Review requested from: ${Array.from(reviewer_persons_set)}`
      })

      const { data: prDiff } = await octokit.rest.pulls.get({
        owner: pr_owner,
        repo: repo,
        pull_number: pr_number,
        mediaType: {
          format: "diff"
        }
      })

      console.log(prDiff.body)

    } else {
      console.log(`I don't care about requesting approvals! Will just check who already approved`)

      // aggregate reviewers from persons and teams
      console.log(`org: ${organization}`)

      const teams_list = await octokit.rest.teams.list({
        ...context.repo,
        org: organization
      });

      for (const team of teams_list.data) {
        console.log(`team list: ${team.slug}`)

        const team_list_obj = await octokit.rest.teams.listMembersInOrg({
          ...context.repo,
          org: organization,
          team_slug: team.slug
        });

        for (const member of team_list_obj.data) {
          if (pr_owner != member!.login) {
            console.log(`team_member: ${member!.login!}`) //debug output
            reviewer_persons_set.add(member!.login)
          }
        }
      }

      //retrieve approvals
      const reviews = await octokit.rest.pulls.listReviews({
        ...context.repo,
        pull_number: payload.pull_request.number
      })
      const approved_users: Set<string> = new Set()
      for (const review of reviews.data) {
        if (review.state === `APPROVED`) {
          approved_users.add(review.user!.login)
          console.log(`Approved: ${review.user!.login} --- ${review.state}`)
        } else {
          approved_users.delete(review.user!.login)
          console.log(`Other state: ${review.user!.login} --- ${review.state}`)
        }
      }

      // check approvals
      const review_gatekeeper = new ReviewGatekeeper(
        config_file_contents as Settings,
        Array.from(approved_users),
        reviewer_persons_set,
        payload.pull_request.user.login
      )

      // The workflow url can be obtained by combining several environment varialbes, as described below:
      // https://docs.github.com/en/actions/reference/environment-variables#default-environment-variables
      core.info(`Setting a status on commit (${sha})`)


      octokit.rest.repos.createCommitStatus({
        ...context.repo,
        sha,
        state: review_gatekeeper.satisfy() ? 'success' : 'failure',
        context: workflow_name,
        target_url: workflow_url,
        description: review_gatekeeper.satisfy()
          ? undefined
          : review_gatekeeper.getMessages().join(' ')
      })

      if (!review_gatekeeper.satisfy()) {
        core.setFailed(review_gatekeeper.getMessages().join(EOL))
        return
      }
    }
  } catch (error) {
    core.setFailed(error.message)
    console.log("error: ", error)
  }
}

run()
