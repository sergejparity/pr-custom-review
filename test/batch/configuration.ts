import { Octokit } from "@octokit/rest"
import nock from "nock"
import YAML from "yaml"

import {
  basePR,
  condition,
  configFileContentsApiPath,
  configFilePath,
  githubApi,
  githubWebsite,
  rulesExamples,
  team,
  team2,
} from "test/constants"
import Logger from "test/logger"

import {
  rulesConfigurations,
  variableNameToActionInputName,
} from "src/constants"
import { runChecks } from "src/core"

describe("Configuration", function () {
  let logger: Logger
  let octokit: Octokit
  let logHistory: string[]

  beforeEach(function () {
    nock.disableNetConnect()
    logHistory = []
    logger = new Logger(logHistory)
    octokit = new Octokit()
    nock(githubWebsite)
      .get(basePR.diff_url.slice(githubWebsite.length))
      .reply(200, condition)
  })

  for (const [missingField, value] of [
    ["name", condition],
    ["condition", condition],
    ["check_type", "diff"],
    ["min_approvals", 1],
  ]) {
    it(`Configuration is invalid if ${missingField} is missing`, async function () {
      nock(githubApi)
        .get(configFileContentsApiPath)
        .reply(200, {
          content: Buffer.from(
            `
            rules:
              - ${missingField === "name" ? "" : `name: ${value}`}
                ${missingField === "condition" ? "" : `condition: ${value}`}
                ${missingField === "check_type" ? "" : `check_type: ${value}`}
                ${
                  missingField === "min_approvals"
                    ? ""
                    : `min_approvals: ${value}`
                }
            `,
          ).toString("base64"),
        })

      expect(
        await runChecks(basePR, octokit, logger, {
          configFilePath,
          locksReviewTeam: team,
          teamLeadsTeam: team2,
        }),
      ).toBe("failure")

      expect(logHistory).toMatchSnapshot()
    })
  }

  it("Configuration is invalid if min_approvals is less than 1", async function () {
    nock(githubApi)
      .get(configFileContentsApiPath)
      .reply(200, {
        content: Buffer.from(
          `
          rules:
            - name: ${condition}
              condition: ${condition}
              check_type: diff
              min_approvals: 0
          `,
        ).toString("base64"),
      })

    expect(
      await runChecks(basePR, octokit, logger, {
        configFilePath,
        locksReviewTeam: team,
        teamLeadsTeam: team2,
      }),
    ).toBe("failure")

    expect(logHistory).toMatchSnapshot()
  })

  for (const name in variableNameToActionInputName) {
    it(`Configuration is invalid if ${name} is empty or missing`, async function () {
      expect(
        await runChecks(basePR, octokit, logger, {
          configFilePath,
          locksReviewTeam: team,
          teamLeadsTeam: team2,
          [name]: "",
        }),
      ).toBe("failure")

      expect(logHistory).toMatchSnapshot()
    })
  }

  for (const { kind, invalidFields } of Object.values(rulesConfigurations)) {
    const goodRule = rulesExamples[kind]

    for (const invalidField of invalidFields) {
      const invalidFieldValidValue = (function () {
        switch (invalidField) {
          case "all":
          case "teams":
          case "users":
          case "any": {
            return []
          }
          case "min_approvals": {
            return 1
          }
          default: {
            const exhaustivenessCheck: never = invalidField
            throw new Error(
              `invalidField is not handled: ${exhaustivenessCheck}`,
            )
          }
        }
      })()

      it(`Rule kind ${kind} does not allow invalid field ${invalidField}`, async function () {
        const badRule = { ...goodRule, [invalidField]: invalidFieldValidValue }

        nock(githubApi)
          .get(configFileContentsApiPath)
          .reply(200, {
            content: Buffer.from(YAML.stringify({ rules: [badRule] })).toString(
              "base64",
            ),
          })

        expect(
          await runChecks(basePR, octokit, logger, {
            configFilePath,
            locksReviewTeam: team,
            teamLeadsTeam: team2,
          }),
        ).toBe("failure")

        expect(logHistory).toMatchSnapshot()
      })
    }
  }

  for (const [kind, exampleRule] of Object.entries(rulesExamples)) {
    for (const [invalidValue, description] of [
      [0, "less than 1"],
      [null, "null"],
    ]) {
      it(`min_approvals is invalid for ${kind} if it is ${description}`, async function () {
        nock(githubApi)
          .get(configFileContentsApiPath)
          .reply(200, {
            content: Buffer.from(
              YAML.stringify({
                rules: [
                  {
                    ...exampleRule,
                    ...("min_approvals" in exampleRule
                      ? { min_approvals: invalidValue }
                      : "all" in exampleRule
                      ? { all: [{ min_approvals: invalidValue }] }
                      : "any" in exampleRule
                      ? { any: [{ min_approvals: invalidValue }] }
                      : {}),
                  },
                ],
              }),
            ).toString("base64"),
          })

        expect(
          await runChecks(basePR, octokit, logger, {
            configFilePath,
            locksReviewTeam: team,
            teamLeadsTeam: team2,
          }),
        ).toBe("failure")

        expect(logHistory).toMatchSnapshot()
      })
    }
  }

  afterEach(function () {
    nock.cleanAll()
    nock.enableNetConnect()
  })
})
