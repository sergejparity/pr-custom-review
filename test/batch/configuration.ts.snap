// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration Configuration is invalid if actionReviewTeam is empty or missing 1`] = `
Array [
  "ERROR:  Action Review Team (action input: action-review-team) should be provided",
]
`;

exports[`Configuration Configuration is invalid if check_type is missing 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].check_type\\" is required] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].check_type\\" is required',
      path: [Array],
      type: 'any.required',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Configuration is invalid if condition is missing 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].condition\\" is required] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].condition\\" is required',
      path: [Array],
      type: 'any.required',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Configuration is invalid if locksReviewTeam is empty or missing 1`] = `
Array [
  "ERROR:  Locks Review Team (action input: locks-review-team) should be provided",
]
`;

exports[`Configuration Configuration is invalid if min_approvals is less than 1 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].min_approvals\\" must be greater than or equal to 1] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].min_approvals\\" must be greater than or equal to 1',
      path: [Array],
      type: 'number.min',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Configuration is invalid if min_approvals is missing 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].name\\" must be a string] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].name\\" must be a string',
      path: [Array],
      type: 'string.base',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Configuration is invalid if name is missing 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].name\\" is required] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].name\\" is required',
      path: [Array],
      type: 'any.required',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Configuration is invalid if teamLeadsTeam is empty or missing 1`] = `
Array [
  "ERROR:  Team Leads Team (action input: team-leads-team) should be provided",
]
`;

exports[`Configuration Rule kind AndRule does not allow invalid field any 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"AndRule\\" because it had the field \\"all\\", but it also has the field \\"any\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind AndRule does not allow invalid field min_approvals 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"min_approvals\\", but it also has the field \\"all\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind AndRule does not allow invalid field teams 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"teams\\", but it also has the field \\"all\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind AndRule does not allow invalid field users 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"users\\", but it also has the field \\"all\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind BasicRule does not allow invalid field all 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"min_approvals\\", but it also has the field \\"all\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind BasicRule does not allow invalid field any 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"min_approvals\\", but it also has the field \\"any\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind OrRule does not allow invalid field all 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"AndRule\\" because it had the field \\"all\\", but it also has the field \\"any\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind OrRule does not allow invalid field min_approvals 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"min_approvals\\", but it also has the field \\"any\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind OrRule does not allow invalid field teams 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"teams\\", but it also has the field \\"any\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind OrRule does not allow invalid field users 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"users\\", but it also has the field \\"any\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration min_approvals is invalid for AndRule if it is less than 1 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].all[0].min_approvals\\" must be greater than or equal to 1] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].all[0].min_approvals\\" must be greater than or equal to 1',
      path: [Array],
      type: 'number.min',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration min_approvals is invalid for AndRule if it is null 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].all[0].min_approvals\\" must be a number] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].all[0].min_approvals\\" must be a number',
      path: [Array],
      type: 'number.base',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration min_approvals is invalid for BasicRule if it is less than 1 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].min_approvals\\" must be greater than or equal to 1] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].min_approvals\\" must be greater than or equal to 1',
      path: [Array],
      type: 'number.min',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration min_approvals is invalid for BasicRule if it is null 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "Matched expression \\"condition\\" of rule \\"condition\\" on diff",
  "ERROR:  Rule \\"condition\\" has invalid min_approvals",
  "{
  name: 'condition',
  condition: 'condition',
  check_type: 'diff',
  min_approvals: null
}",
]
`;

exports[`Configuration min_approvals is invalid for OrRule if it is less than 1 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].any[0].min_approvals\\" must be greater than or equal to 1] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].any[0].min_approvals\\" must be greater than or equal to 1',
      path: [Array],
      type: 'number.min',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration min_approvals is invalid for OrRule if it is null 1`] = `
Array [
  "Changed files Set(1) { 'condition' }",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].any[0].min_approvals\\" must be a number] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].any[0].min_approvals\\" must be a number',
      path: [Array],
      type: 'number.base',
      context: [Object]
    }
  ]
}",
]
`;
